File: src/tools/namespace.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { GraphQLClient } from 'graphql-request';
import { z } from 'zod';
import {
  NamespaceListDocument,
  NamespaceGetDocument,
  NamespaceCreateDocument,
  NamespaceDeleteDocument,
  type NamespaceListQuery,
  type NamespaceGetQuery,
  type NamespaceCreateMutation,
} from '../generated/graphql.js';

export function registerNamespaceTools(server: McpServer, client: GraphQLClient): void {
  server.registerTool(
    'nexaa_namespace_list',
    { description: 'List all namespaces in your Nexaa account' },
    async () => {
      const data = await client.request<NamespaceListQuery>(NamespaceListDocument);
      return { content: [{ type: 'text', text: JSON.stringify(data.namespaces, null, 2) }] };
    },
  );

  server.registerTool(
    'nexaa_namespace_get',
    {
      description: 'Get a namespace by name',
      inputSchema: { name: z.string().describe('Namespace name') },
    },
    async ({ name }) => {
      const data = await client.request<NamespaceGetQuery>(NamespaceGetDocument, { name });
      return { content: [{ type: 'text', text: JSON.stringify(data.namespace, null, 2) }] };
    },
  );

  server.registerTool(
    'nexaa_namespace_create',
    {
      description: 'Create a new namespace',
      inputSchema: {
        name: z.string().describe('Namespace name'),
        description: z.string().optional().describe('Optional description'),
      },
    },
    async ({ name, description }) => {
      const data = await client.request<NamespaceCreateMutation>(NamespaceCreateDocument, {
        input: { name, description },
      });
      return { content: [{ type: 'text', text: JSON.stringify(data.namespaceCreate, null, 2) }] };
    },
  );

  server.registerTool(
    'nexaa_namespace_delete',
    {
      description:
        'Delete a namespace. WARNING: This will delete the namespace and all its resources. Requires confirm: true.',
      inputSchema: {
        name: z.string().describe('Namespace name to delete'),
        confirm: z.boolean().describe('Must be set to true to confirm deletion'),
      },
    },
    async ({ name, confirm }) => {
      if (!confirm) {
        return {
          content: [
            {
              type: 'text',
              text: 'Deletion cancelled: confirm must be true to delete a namespace.',
            },
          ],
        };
      }
      await client.request(NamespaceDeleteDocument, { name });
      return { content: [{ type: 'text', text: `Namespace "${name}" deleted.` }] };
    },
  );
}

File: tools/namespace.go
package tools

import (
	"context"
	"fmt"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
	"github.com/nexaa-cloud/nexaa-cli/api"
)

func RegisterNamespace(s *server.MCPServer) {
	s.AddTool(
		mcp.NewTool("nexaa_namespace_list",
			mcp.WithDescription("List all namespaces in your Nexaa account"),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			result, err := withAPI(func(c *api.Client) ([]api.NamespaceResult, error) {
				return c.NamespacesList()
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_namespace_get",
			mcp.WithDescription("Get a namespace by name"),
			mcp.WithString("name", mcp.Required(), mcp.Description("Namespace name")),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			name := req.GetString("name", "")
			result, err := withAPI(func(c *api.Client) (api.NamespaceResult, error) {
				return c.NamespaceListByName(name)
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_namespace_create",
			mcp.WithDescription("Create a new namespace"),
			mcp.WithString("name", mcp.Required(), mcp.Description("Namespace name")),
			mcp.WithString("description", mcp.Description("Optional description")),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			name := req.GetString("name", "")
			desc := req.GetString("description", "")
			input := api.NamespaceCreateInput{Name: name}
			if desc != "" {
				input.Description = &desc
			}
			result, err := withAPI(func(c *api.Client) (api.NamespaceResult, error) {
				return c.NamespaceCreate(input)
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_namespace_delete",
			mcp.WithDescription("Delete a namespace. WARNING: This will delete the namespace and all its resources. Requires confirm: true."),
			mcp.WithString("name", mcp.Required(), mcp.Description("Namespace name to delete")),
			mcp.WithBoolean("confirm", mcp.Required(), mcp.Description("Must be set to true to confirm deletion")),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			name := req.GetString("name", "")
			if ok, cancelled := confirmDelete(req, "namespace"); !ok {
				return cancelled, nil
			}
			_, err := withAPI(func(c *api.Client) (bool, error) {
				return c.NamespaceDelete(name)
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return mcp.NewToolResultText(fmt.Sprintf("Namespace %q deleted.", name)), nil
		},
	)
}
